memory.go

```diff
@@ -0,0 +1,227 @@
+		return nil, 0, errors.New("this Go runtime has a moving garbage collector")
+	}
+
+	// Request at least a full page from the runtime, otherwise the allocated span
+	// is likely not page-aligned, risking mapping over objects on another page.
+	// Since we use MAP_FIXED, the starting address of the mapping must be
+	// page-aligned on most architectures anyway.
+	size = internal.Align(size, os.Getpagesize())
+
+	// Get the address of the backing array of the slice and check if it's
+	// page-aligned to make sure it works with MAP_FIXED.
+	ptr := unsafe.Pointer(unsafe.SliceData(make([]byte, size)))
+	addr := int(uintptr(ptr))
+	if internal.Align(addr, os.Getpagesize()) != addr {
+		return nil, 0, fmt.Errorf("allocated memory is not page-aligned: %d", addr)
+	}
+
+	return ptr, size, nil
+}
+
+// unmap returns a function that takes a pointer to a memory-mapped region on
+// the Go heap. The function undoes any mappings and discards the span's
+// contents.
+//
+// Used as a finalizer in [newMemory], split off into a separate function for
+// testing and to avoid accidentally closing over the unsafe.Pointer to the
+// memory region, which would cause a cyclical reference.
+//
+// The resulting function panics if the mmap operation returns an error, since
+// it would mean the integrity of the Go heap is compromised.
+func unmap(size int) func(*byte) {
+	return func(a *byte) {
+		// Create another mapping at the same address to undo the original mapping.
+		// This will cause the kernel to repair the slab since we're using the same
+		// protection mode and flags as the original mapping for the Go heap.
+		//
+		// Address           Kbytes     RSS   Dirty Mode  Mapping
+		// 000000c000000000    4096     884     884 rw--- [ anon ]
+		//
+		// Using munmap here would leave an unmapped hole in the heap, compromising
+		// its integrity.
+		//
+		// MmapPtr allocates another unsafe.Pointer at the same address. Even though
+		// we discard it here, it may temporarily resurrect the backing array and
+		// delay its collection to the next GC cycle.
+		_, err := unix.MmapPtr(-1, 0, unsafe.Pointer(a), uintptr(size),
+			unix.PROT_READ|unix.PROT_WRITE,
+			unix.MAP_PRIVATE|unix.MAP_FIXED|unix.MAP_ANONYMOUS)
+		if err != nil {
+			panic(fmt.Errorf("undoing bpf map memory mapping: %w", err))
+		}
+	}
+}
+
+// Size returns the size of the memory-mapped region in bytes.
+func (mm *Memory) Size() int {
+	return len(mm.b)
+}
+
+// Readonly returns true if the memory-mapped region is read-only.
+func (mm *Memory) Readonly() bool {
+	return mm.ro
+}
+
+// ReadAt implements [io.ReaderAt]. Useful for creating a new [io.OffsetWriter].
+func (mm *Memory) ReadAt(p []byte, off int64) (int, error) {
+	if mm.b == nil {
+		return 0, fmt.Errorf("memory-mapped region closed")
+	}
+
+	if p == nil {
+		return 0, fmt.Errorf("input buffer p is nil")
+	}
+
+	if off < 0 || off >= int64(len(mm.b)) {
+		return 0, fmt.Errorf("read offset out of range")
+	}
+
+	n := copy(p, mm.b[off:])
```

Comment:
It would be nice if `ReadAt` copied properly aligned chunks in word-size units. Imagine I copy a `struct` with counters that is concurrently updated by bpf program. Should the `copy` decide to go one byte at a time, values can be seen that never were.

I suppose builtin `copy` is probably doing exactly that, albeit non-contractual. (Very minor and can wait until later.)

Reply:
> Should the copy decide to go one byte at a time, values can be seen that never were.

Without using atomics, the memory model(s) of the architectures we run on don't make any guarantees about the consistency of values under concurrent access. To my limited understanding, on x64, the smallest unit of memory that can be shared between cores/threads is a cache line, representing 8 bytes. If kernel space is writing and user space is reading, these cache lines can be stale by the time the reader fetches them. Afaik, this is only a real issue if a value spans multiple cache lines, like a float/complex128 or a buffer of arbitrary length. In general, struct alignment rules make sure values stay contained within a single cache line.

I'm not sure what change you're proposing to our use of copy(). I haven't looked at copy's implementation in detail, but ultimately we have very little control over how the compiler chooses to shuffle memory acesses around through several optimization passes, unless copy() contains some kind of memory barrier. (unlikely, need to check) If you're suggesting to batch our calls to copy(), that sounds ineffectual to me.

If you're worried about reading a uint64 that has 4 out of its 8 bytes gone stale, afaik that's not possible as long as you stay away from packed structs. Batching copy() won't save your bacon here either. If you need synchronization, pair `sync/atomic` accesses in Go with `__sync_` in C.

Reply:
> If you're worried about reading a uint64 that has 4 out of its 8 bytes gone stale, afaik that's not possible as long as you stay away from packed structs.

This is exactly what I'd like to prevent (an **aligned** `uint64` appearing as if it was partially updated when fetched via `ReadAt`).

If copy goes one byte at a time, it could copy the first 4 bytes, get preempted, and get back to copying the rest later. That would be a conformant `copy` implementation, albeit not a very fast one.

> I'm not sure what change you're proposing to our use of copy().

No change proposed :) We might need to implement our own copy if it turns out that builtin misbehaves on some platforms. [amd64](https://github.com/golang/go/blob/master/src/runtime/memmove_amd64.s#L35) seems to be ok.

Reply:
The more I think about it, the more I'm considering removing byte-wise reads/writes from Variable altogether. I think `Memory` still needs them to be able to initialize slabs of memory after creating an arena map, since it's not practical to use the syscall map API for it if it works at all.

I hadn't considered the preemption case for some reason. Newer versions of the runtime can preempt after any instruction aside from some 'critical sections', which I haven't found a definition for. Disabling preemption within a function boundary (https://github.com/golang/go/issues/65874) is currently not possible, but even if we manage to prevent runtime preemption, there's still the OS scheduler.

For shared memory accesses to be sound, they need to occur in a single instruction.

> [amd64](https://github.com/golang/go/blob/master/src/runtime/memmove_amd64.s#L35) seems to be ok.

What's your conclusion exactly from reading this memmove implementation?

Reply:
> What's your conclusion exactly from reading this memmove implementation?

Memory is not read a byte at a time unless absolutely necessary. Implementation picks a transfer unit based on the total size. Could be 8 byte at a time (general-purpose registers), could be 16 (SSE).

It doesn't care whether source or destination is aligned. Assuming transfer size of 8 was picked: it will do a 8 byte load at `+0`, another load at `+8`, etc. Values are consistent as long as loads don't cross a cache-line boundary. Of course, there are no ordering guarantees, but we never see partially updated values.

An interesting case is when the total size is NOT a multiple of the transfer size ([`move_9through16`](https://github.com/golang/go/blob/master/src/runtime/memmove_amd64.s#L192)):
```
| 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 |
|***|***|***|***|***|***|***|***|   |   | <-- first load
|   |   |***|***|***|***|***|***|***|***| <-- second load
                                |
                                | could be a cache line boundary
| A | A | b | b | b | b | b | b | c | c | cells annotated with "epochs"
```
Some loads are performed at `+8`, etc. relative to block **end**, creating some overlap.

For the purposes of this discourse, let's assume that the start is 8-byte aligned. Then there *might* be a cache line boundary breaking our block into 8 and 2 byte chunks. The second load *might* cross a cache line boundary. I believe it is specified somewhere; CPU fetches 2 cache lines separately and internally combines the data. So after copy completes, the first 2 bytes are from epoch `A`, followed by 6 bytes from epoch `b`, followed by 2 bytes from epoch `c`.

C structure layout rules defines structure alignment as the max alignment between the members. Structure size in turn is a multiple of the structure alignment. Should the total size of the structure be 10, it indicates that the struct alignment was 2 or 1. There were no members larger than `u16`. Hence no member received a partially updated value. (Note: so far we assumed 8 byte aligned block. We should consider +2, +4, and +6 misaligned blocks for the complete proof.)

To summarise, it seems to be ok to use `copy` to capture structures from concurrently modified memory on amd64. There are no relative ordering guarantees, but we never see a partially updated field.

> I hadn't considered the preemption case for some reason. [...] For shared memory accesses to be sound, they need to occur in a single instruction.

Well, we can't guarantee that we obtain a consistent "snapshot" of a variable, but a weaker guarantee of "no partially updated fields" is also useful in practice.

> The more I think about it, the more I'm considering removing byte-wise reads/writes from Variable altogether.

It's reasonable to be somewhat concerned :) That's why I wanted to learn more about `copy` implementation. Note: Linux is facing a similar challenge, when doing `bpf_map_lookup_elem` through the syscall, value is memory-copied w/o obtaining any locks.

I think we should keep reads for read-only variables, the alternative is to expose pointers that will fault on write (AFAIK go can't recover from SIGBUS).

Reads/writes are also nice to have for the symmetry with maps api; objects used as K/V can define their own custom serialisation methods.

Reply:
Thank you for the analysis, that was really helpful! I have a few more points to add before we can reach consensus:

> but a weaker guarantee of "no partially updated fields" is also useful in practice.

Of course, consistency across fields of a struct can never be guaranteed without some form of synchronization, but if there's a chance for either side to see a partially-written value, that's a problem. Fortunately, struct alignment rules are on our side here. I'll make sure to document that packed structs should be avoided at all costs and summarize the hazards we flagged in this discussion. Now we know how memmove works at a high level, I'm more comfortable leaving Variable.Get() and .Set() in the proposal.

I briefly looked at the arm and arm64 implementations as well, and both of them unsurprisingly try to copy as much data in a single instruction as possible, with the pathological case being unaligned or odd-sized (e.g. copying 3 bytes) as expected.

It looks like arm32 can fill 32 bytes of memory into 8 4-byte registers (`MOVM.DB.W (FROM), [R0-R7]`) in a single instruction. My only doubt here is that this seems to be limited to 32 and 16 bytes, notably missing the 8-byte case for the common u64 and bpf pointers. These seem to be handled using 2 4-byte chunks in a loop, which could be prone to preemption.

I'm not sure if we should let this force our hand, as I'm not even sure how much of bpf actually works on arm32 nowadays. When I tried on an old Pi a few years ago, it simply didn't, but I've heard some people use it.

> the alternative is to expose pointers that will fault on write

I don't think we should allow that, it's too big of a footgun. Someone changes a C global to a const and their application gets terminated. A Go program failing like this is highly surprising, so I'd rather not allow it. For consts, Variable.Get() works fine and is safe, and I'd expect most consts to only be written to from user space; after all the application itself is the source of truth for these.

---

tl;dr - the current proposal is sufficient, modulo documentation around memory alignment? The manual still needs to be written anyway, so I can keep it in mind from the start.

Reply:
Please feel free to mark as resolved.